src/viam/sdk/common/client_helper.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <grpcpp/client_context.h>
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+
+namespace viam {
+namespace sdk {
+
+template <typename ClientType, typename StubType, typename RequestType, typename ResponseType>
+class ClientHelper {
+    static void default_rsc_(RequestType&) {}
+    static void default_rhc_(const ResponseType&) {}
+    static void default_ehc_(const ::grpc::Status& status) {
+        throw std::runtime_error(status.error_message());
+    }
+
+   public:
+    using PFn = ::grpc::Status (StubType::*)(::grpc::ClientContext*,
+                                             const RequestType&,
+                                             ResponseType*);
+    explicit ClientHelper(ClientType* client, StubType* stub, PFn pfn)
+        : client_(client), stub_(stub), pfn_(pfn) {}
+
+    ClientHelper& with(const AttributeMap& extra) {
+        return with(extra, default_rsc_);
+    }
+
+    template <typename RequestSetupCallable>
+    ClientHelper& with(RequestSetupCallable&& rsc) {
+        std::forward<RequestSetupCallable>(rsc)(request_);
+        return *this;
+    }
+
+    template <typename RequestSetupCallable>
+    ClientHelper& with(const AttributeMap& extra, RequestSetupCallable&& rsc) {
+        *request_.mutable_extra() = map_to_struct(extra);
+        return with(std::forward<RequestSetupCallable>(rsc));
+    }
+
+    template <typename ResponseHandlerCallable = decltype(default_rhc_)>
+    auto invoke(ResponseHandlerCallable&& rhc = default_rhc_) {
```

Comment:
Yeah FWIW I appreciate a named `invoke` method over overloading `operator()` because an `invoke` with an rhc and an ehc may be pretty tough to read. The `ServiceHelper` _does_ do the opposite, but it kind of makes sense that the client is the one "invoking" the gRPC endpoint.